elfo-logger/src/line_buffer.rs

```diff
@@ -0,0 +1,256 @@
+        need_to_erase -= payload_part;
+        let fields_part = fields_len.min(need_to_erase);
+        need_to_erase -= fields_part;
+        let meta_part = meta_len.min(need_to_erase);
+
+        self.0.buf.payload.truncate(payload_len - payload_part);
+        self.0.buf.fields.truncate(fields_len - fields_part);
+        self.0
+            .buf
+            .buffer
+            .truncate(self.0.buf.buffer.len() - meta_part);
+    }
+}
+
+impl<'a> Drop for CurrentLine<'a> {
+    fn drop(&mut self) {
+        self.probe_size_limit();
+
+        {
+            let buffer = &mut self.0.buf.buffer;
+            buffer.push_str(&self.0.buf.payload);
+            buffer.push_str(&self.0.buf.fields);
+            buffer.push('\n');
+        }
+    }
+}
+
+// DirectWrite
+
+#[derive(Debug)]
+pub(crate) struct DirectWrite<'a>(Repr<'a>);
+
+impl<'a> Drop for DirectWrite<'a> {
+    fn drop(&mut self) {
+        self.0.buf.buffer.push('\n');
+    }
+}
+
+impl Line for DirectWrite<'_> {
+    fn discard(self) {
+        self.0.buf.buffer.truncate(self.0.pre_start_buffer_size);
+
+        mem::forget(self);
+    }
+
+    fn total_wrote(&self) -> usize {
+        self.0.buf.buffer.len() - self.0.pre_start_buffer_size
+    }
+
+    fn meta_mut(&mut self) -> &mut String {
+        &mut self.0.buf.buffer
+    }
+
+    fn payload_mut(&mut self) -> &mut String {
+        &mut self.0.buf.buffer
+    }
+
+    fn fields_mut(&mut self) -> &mut String {
+        &mut self.0.buf.buffer
+    }
+}
+
+// LineBuffer
+
+#[derive(Debug)]
+pub(crate) struct LineBuffer {
+    // Buffer will contain lines and partially written meta-info, such as `<timestamp> <level>
+    // [<trace_id>]` `payload` and `fields` will be merged as soon as [`CurrentLine`] will be
+    // finished
+    pub(crate) buffer: String,
+
+    payload: String,
+    fields: String,
+
+    pub(crate) max_line_size: usize,
+}
+
+impl Default for LineBuffer {
+    fn default() -> Self {
+        Self::new(usize::MAX)
+    }
+}
+
+impl LineBuffer {
+    /// Discards current buffers
+    pub(crate) fn clear(&mut self) {
+        self.buffer.clear();
+        self.payload.clear();
+        self.fields.clear();
+    }
+
+    fn create_repr(&mut self) -> Repr<'_> {
+        let size = self.buffer.len();
+        Repr {
+            buf: self,
+            pre_start_buffer_size: size,
+        }
+    }
+
+    pub(crate) fn direct_write(&mut self) -> DirectWrite<'_> {
+        DirectWrite(self.create_repr())
+    }
+
+    pub(crate) fn new_line(&mut self) -> CurrentLine<'_> {
+        CurrentLine(self.create_repr())
+    }
+
+    pub(crate) fn with_buffer_capacity(capacity: usize, max_line_size: usize) -> Self {
+        Self {
+            buffer: String::with_capacity(capacity),
+            payload: String::new(),
+            fields: String::new(),
+            max_line_size,
+        }
+    }
+
+    pub(crate) const fn new(max_line_size: usize) -> Self {
```

Comment:
It's actually used only in tests, so I see no reason to provide both (`with_capacity`) and `new`/`Default`.

If you still want to leave all these constructors, let's, at least, reduce initialization code by rewriting `with_capacity` in the following way:
```rust
let mut this = Self::new(max_line_size);
this.buffer.reserve(capacity);
this
```